packages/grafana-flamegraph/src/FlameNodeGraph/edgeUtils.test.ts

```diff
@@ -0,0 +1,125 @@
+
+  it('doesnt create new edge if its not needed', () => {
+    const { root, nodes, edges } = getDefaultGraph();
+    delete nodes['1'];
+
+    const newEdges = createResidualEdges(root, nodes, edges);
+    expect(newEdges).toEqual(edges);
+  });
+
+  it('creates residual edge if we remove middle node', () => {
+    const { root, nodes, edges } = getDefaultGraph();
+    delete nodes['4'];
+
+    const newEdges = createResidualEdges(root, nodes, edges);
+    expect(edgesToString(newEdges)).toEqual(['1-3', '2-4', '2-5', '4-5']);
+    expect(newEdges['2-5']).toMatchObject({ residual: true });
+  });
+});
+
+describe('removeRedundantEdges', () => {
+  it('it should remove redundant edges', () => {
+    // At the start we have graph like this:
+    //  0   1
+    //  |\ /
+    //  2 3
+    //  \ |
+    //   4
+
+    const allSortedNodes: GraphNode[] = [
+      { label: '0', self: 5, value: 5, parents: [], children: [] },
+      { label: '1', self: 6, value: 6, parents: [], children: [] },
+      { label: '2', self: 3, value: 3, parents: [], children: [] },
+      { label: '3', self: 6, value: 6, parents: [], children: [] },
+      { label: '4', self: 3, value: 3, parents: [], children: [] },
+    ];
+
+    const edges: GraphEdges = {
+      '0-2': { from: allSortedNodes[0], to: allSortedNodes[2], weight: 1, residual: false },
+      '0-3': { from: allSortedNodes[0], to: allSortedNodes[3], weight: 1, residual: false },
+      '1-3': { from: allSortedNodes[1], to: allSortedNodes[3], weight: 1, residual: false },
+      '2-4': { from: allSortedNodes[2], to: allSortedNodes[4], weight: 1, residual: false },
+      '3-4': { from: allSortedNodes[3], to: allSortedNodes[4], weight: 1, residual: false },
+    };
+
+    allSortedNodes[0].children.push(edges['0-2'], edges['0-3']);
+    allSortedNodes[1].children.push(edges['1-3']);
+    allSortedNodes[2].parents.push(edges['0-2']);
+    allSortedNodes[2].children.push(edges['2-4']);
+    allSortedNodes[3].parents.push(edges['0-3'], edges['1-3']);
+    allSortedNodes[3].children.push(edges['3-4']);
+    allSortedNodes[4].parents.push(edges['2-4'], edges['3-4']);
+
+    // By trimming we removed 3:
+    //  0   1
+    //  |\ /
+    //  2 x
+    //  \ |
+    //   4
+
+    const keptNodes: GraphNodes = {
+      0: allSortedNodes[0],
+      1: allSortedNodes[1],
+      2: allSortedNodes[2],
+      4: allSortedNodes[4],
+    };
+
+    // This should then create residual edges:
+    //  0   1
+    //  |\  |
+    //  2 | |
+    //  \ | /
+    //   \|/
+    //    4
+    // Where 0-4 is redundant while 1-4 is kept. We should also remove any edge referencing the removed 3 node.
+    edges['0-4'] = { from: allSortedNodes[0], to: allSortedNodes[4], weight: 1, residual: true };
+    edges['1-4'] = { from: allSortedNodes[1], to: allSortedNodes[4], weight: 1, residual: true };
+
+    const newEdges = removeRedundantEdges(allSortedNodes, keptNodes, edges, 0);
+    expect(edgesToString(newEdges)).toEqual(['0-2', '1-4', '2-4']);
```

Comment:
The code is producing the in my eyes correct result. 

You want to show that function 0 has called 4 (directly, or with nodes in between that have been removed), rather than add the value of the mesurement to the one were function 0 calls 2 which then calls 4 

```suggestion
    expect(edgesToString(newEdges)).toEqual(['0-2', '0-4', '1-4', '2-4']);
```

Reply:
@simonswine I think you are right but I am trying to figure out why this function is written like it is:
```typescript
function isRedundantEdge(edge: GraphEdge) {
  const seen: Record<string, boolean> = {};
  const queue = [edge.to];

  while (queue.length > 0) {
    const node = queue.shift()!;

    // loop over all edges that go into this node
    for (const parentEdge of node.parents) {
      if (edge === parentEdge || seen[parentEdge.from.label]) {
        // skip if we are looking at the same edge, or we have already seen this parent node
        continue;
      }

      if (parentEdge.from === edge.from) {
        // we found edge that has the same from - to nodes as the edge we are testing.
        return true;
      }
      seen[parentEdge.from.label] = true;
      queue.push(parentEdge.from);
    }
  }
  return false;
}

```

I mean it seems to just check if there is any duplicate edges. Which means edge that has the same `from` and `to` node. Like that seems to be it so it could just check all the parent edges of it's `to` node and check it's `from` node and that's it. I am not sure why it does a traversal with `queue.push(parentEdge.from);`. In case of graph like this:
```
0   1
|\  |
2 | |
\ | /
 \|/
  4
```
Testing 2-4 means checking 0-4 and 1-4 which obviously don't have the same nodes but then it goes to node 2 and checks its edges but because we have already seen 0 we skip. That seems unnecessary to me unless I don't understand some other parts of this.

Reply:
But that should not even happen because we pass only a map of edges where key is `<from>-<to>` and so you cannot even create such duplicates in such map, because you would just rewrite the key.

Reply:
Ok it does not actually do that. the code marks nodes it sees except for the current edge nodes so if I have edges `1-2` `1-3` and `3-2`, if I test `1-2` it will be redundant because I can traverse to `1` through `3`. Why doesn't do this in test is this weird part of code:
https://github.com/grafana/grafana/blob/1ef4d495995ab71d26e78e7632790236e61cb8a2/packages/grafana-flamegraph/src/FlameNodeGraph/edgeUtils.ts#L100

Mind that this is a `break` not `continue`. Which means is we have one edge that is residual all other edges are skipped. This is dependent on the ordering which is done by weight. Again no idea if this is by design or not.

Reply:
When testing on some real world flamegraph it seems to remove a residual edge, that has high weight but then it also trims all the other edges coming from a node because of the weightCutoff here:
https://github.com/grafana/grafana/blob/cea33062232ea3764400b9f270ad73bbbe120054/packages/grafana-flamegraph/src/FlameNodeGraph/edgeUtils.ts#L113

This leaves the node stranded and again no idea if that makes sense or not.